### PR TITLE
ci: Provide example with caching

### PR DESCRIPTION
**Pull request by FlixCoder**

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -24,22 +24,27 @@ jobs:
...
+        run: |
+          cargo +${NIGHTLY_VERSION} fmt -- --check
+          cargo lints clippy --workspace --all-targets -- -D warnings
+          cargo llvm-cov nextest --workspace --ignore-filename-regex "cargo_home.*"
+          cargo +${NIGHTLY_VERSION} test --doc --workspace
+          cargo +${NIGHTLY_VERSION} udeps
+
+      - name: Build
+        shell: bash
+        run: cargo build --release
```
Comment (emgrav) [inline .github/workflows/rust.yml:50]: Does it make sense to do a release build in this workflow? We don't use any caching in the docker build job, and either way they will run concurrently and would probably not benefit from caching.

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -24,22 +24,27 @@ jobs:
...
-    - name: Build
-      run: cargo build --verbose --release
+      - name: Check
+        shell: bash
+        run: |
+          cargo +${NIGHTLY_VERSION} fmt -- --check
+          cargo lints clippy --workspace --all-targets -- -D warnings
+          cargo llvm-cov nextest --workspace --ignore-filename-regex "cargo_home.*"
+          cargo +${NIGHTLY_VERSION} test --doc --workspace
+          cargo +${NIGHTLY_VERSION} udeps
```
Comment (emgrav) [inline .github/workflows/rust.yml:46]: Did we want to split checking, linting and testing into different jobs or steps here?

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -24,22 +24,27 @@ jobs:
...
-        gitlab_user: ${{ secrets.GITLAB_USER }}
-        gitlab_pass: ${{ secrets.GITLAB_PASS }}
-        additional_packages: ${{ env.ADDITIONAL_PACKAGES }}
+      - uses: actions/checkout@v3
+      
+      - uses: ./.github/actions/rust-prepare
+        with:
+          gitlab_ssh: ${{ secrets.CI_SSH_PRIVATE_KEY}}
+          gitlab_user: ${{ secrets.GITLAB_USER }}
+          gitlab_pass: ${{ secrets.GITLAB_PASS }}
```
Comment (emgrav) [inline .github/workflows/rust.yml:33]: If we update all the git dependencies across our repos, we could probably get rid of the gitlab keys and such here, right?

**Comment by emgrav** [comment]
We might also want to add a step to build the crate docs and upload them to github pages using https://github.com/actions/upload-pages-artifact and https://github.com/actions/deploy-pages

Github pages support access control these days, so this can also be done for private repos, so everything should work more or less exactly like it did on gitlab.

File: .github/workflows/rust.yml (hunk 3 above)
Comment (FlixCoder) [reply .github/workflows/rust.yml:33]: I think we should wait until that is done everywhere :D

**Comment by FlixCoder** [comment]
> We might also want to add a step to build the crate docs and upload them to github pages using https://github.com/actions/upload-pages-artifact and https://github.com/actions/deploy-pages
> 
> Github pages support access control these days, so this can also be done for private repos, meaning everything should work more or less exactly like it did on gitlab.

That would be very nice to have, yeah

**Comment by FlixCoder** [comment]
Could also try to add the typos-cli ^^

File: .github/workflows/rust.yml (hunk 2 above)
Comment (FlixCoder) [reply .github/workflows/rust.yml:46]: That's the question, do we want this or not? :D It has advantages and disadvantages

File: .github/workflows/rust.yml (hunk 1 above)
Comment (FlixCoder) [reply .github/workflows/rust.yml:50]: No it was just there, but I can probably remove it

**Comment by emgrav** [comment]
> Could also try to add the typos-cli ^^

The authors of typos-cli actually even provide an action, so this would be super easy: https://github.com/crate-ci/typos/blob/master/docs/github-action.md

**Review by emgrav — APPROVED** [review]
Approved as-is, as we will come back to this later when famedly/product-management#1271 is closed and will need to make further changes then anyway.

**Pull request merged**